mlflow/utils/autologging_utils.py

```diff
@@ -9,3 +11,62 @@ def try_mlflow_log(fn, *args, **kwargs):
         fn(*args, **kwargs)
     except Exception as e:  # pylint: disable=broad-except
         warnings.warn("Logging to MLflow failed: " + str(e), stacklevel=2)
+
+
+def param_logger(fn, args=[], kwargs=None, unlogged=[]):
+    """
+    Log parameters explicitly passed to a function.
+    :param fn: function whose parameters are to be logged
+    :param args: arguments explicitly passed into fn
+    :param kwargs: kwargs explicitly passed into fn
+    :param unlogged: parameters not to be logged
+    :return: None
+    """
+    # Names of all parameters for the function
+    all_param_names = inspect.getargspec(fn)[0]
+
+    # Default values of all parameters with default values. Has length of n, and corresponds
+    # to values of last n elements in all_param_names
+    all_default_values = inspect.getargspec(fn)[3]
+
+    # Checking if default values are present for logging. Known bug that getargspec will return an
+    # empty argspec for certain functions, despite the functions having an argspec.
+    if all_default_values:
+        # Removing the first len(args) elements from all_param_names - these are values already
+        # passed in explicitly by the user and don't need to be logged with default values.
+        kwargs_and_default_names = all_param_names[len(args):]
+
+        # If there are more parameter names than default values left, we know that the parameters
+        # not covered by the default values are passed in as kwargs (assuming all non-default
+        # parameters are passed to the function)
+        if len(kwargs_and_default_names) > len(all_default_values):
+            kwargs_and_default_names = kwargs_and_default_names[len(kwargs_and_default_names)
+                                                                - len(all_default_values):]
+        # Otherwise, if there are more default values than parameter names, we know that some of the
+        # parameters with default values have been entered by the user in args
+        elif len(kwargs_and_default_names) < len(all_default_values):
```

Comment:
How do we know that we can simply exclude the last `kwargs_and_default_names` default values (i.e. that these are the values that correspond to user-specified params)?

Reply:
In these cases we are excluding the first elements of the longer list, not the last. Also, the length of the elements excluded is the difference between `all_default_values` and `kwargs_and_default_values`. In the `elif` case (which I'm assuming you're talking about), we can exclude the first default values since parameters with default arguments must come after parameters without default arguments. At this point of the code, we've already removed the parameter names which have been covered by arguments explicitly passed in by the user in `kwargs_and_default_values`, and if we don't have a 1:1 match of parameter names to values then the first `len(all_default_values)-len(kwargs_and_default_values)` elements of `all_default_values` are known to be covered by user-passed params.